crates/wasmparser/src/validator/component_types.rs

```diff
@@ -1033,6 +1173,46 @@ pub struct RecordType {
     /// The map of record fields.
     pub fields: IndexMap<KebabString, ComponentValType>,
 }
+impl RecordType {
+    fn lower_gc(
+        &self,
+        types: &TypeList,
+        abi: Abi,
+        options: &CanonicalOptions,
+        offset: usize,
+        core: ArgOrField,
+    ) -> Result<()> {
+        let id = match core.as_concrete_ref() {
+            Some(id) => id,
+            None => bail!(
+                offset,
+                "expected to lower component `record` type to core `(ref null? (struct ...))`, \
+                 but found `{core}`",
+            ),
+        };
+
+        match &types[id].composite_type.inner {
+            CompositeInnerType::Struct(ty) => {
+                ensure!(
+                    ty.fields.len() == self.fields.len(),
+                    offset,
+                    "core `struct` has {} fields, but component `record` has {} fields",
+                    ty.fields.len(),
+                    self.fields.len(),
+                );
+                for (core, comp) in ty.fields.iter().zip(self.fields.values()) {
+                    comp.lower_gc(types, abi, options, offset, core.element_type.into())?;
+                }
+                Ok(())
+            }
+            other => bail!(
+                offset,
+                "expected to lower component `record` type to core `(ref null? (struct ...))`, \
+                 but found `{other}`",
+            ),
+        }
+    }
+}
```

Comment:
One thing I've done a lot with canonical ABI things is that there's one helper for packed things and one helper for variant things. For example records/tuples [go here](https://github.com/bytecodealliance/wasmtime/blob/c213302828268411e90a3ea454d42ae254c793e2/crates/environ/src/component/types.rs#L671-L687) while variant/option/result all go [here](https://github.com/bytecodealliance/wasmtime/blob/c213302828268411e90a3ea454d42ae254c793e2/crates/environ/src/component/types.rs#L757-L794). That'll help cut down on duplication between records/tuples here and, in the future, cut down on duplication between variants/option/result.